addon/-private/attr.ts

```diff
@@ -135,22 +132,51 @@ export default function attr(type?: string | AttrOptions, options?: AttrOptions)
           );
         }
       }
-      let internalModel = this._internalModel;
-      if (hasValue(internalModel, key)) {
-        return internalModel.getAttributeValue(key);
+
+      let recordData = getRecordDataFor(this);
+      if (recordData.hasAttr(key)) {
+        return recordData.getAttr(key);
       } else {
         return getDefaultValue(this, options, key);
       }
     },
     set(key, value) {
       if (DEBUG) {
-        if (['_internalModel', 'recordData', 'currentState'].indexOf(key) !== -1) {
+        if (
+          this instanceof Model &&
+          ['_internalModel', 'recordData', 'currentState'].indexOf(key) !== -1
+        ) {
           throw new Error(
             `'${key}' is a reserved property name on instances of classes extending Model. Please choose a different property name for your attr on ${this.constructor.toString()}`
           );
         }
       }
-      return this._internalModel.setDirtyAttribute(key, value);
+      let recordData = getRecordDataFor(this);
+
+      if (this instanceof Model) {
```

Comment:
This branch seems long/complex enough to extract into a function

Reply:
👍